### PR TITLE
Fix shaky scrolling of LazyColumn when the items are of varying size

### LINKED ISSUE #2337: Scrolling is shaky when LazyColumn elements have varying height
SSCCE (taken from https://github.com/JetBrains/compose-jb/tree/master/tutorials/Desktop_Components#lazy-scrollable-components-with-scrollbar ; changed minimally to introduce variance in LazyColumn element's height):

```kotlin
import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application
import androidx.compose.ui.window.rememberWindowState

fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        title = "Scrollbars",
        state = rememberWindowState(width = 250.dp, height = 400.dp)
    ) {
        LazyScrollable()
    }
}

@Composable
fun LazyScrollable() {
    Box(
        modifier = Modifier.fillMaxSize()
            .background(color = Color(180, 180, 180))
            .padding(10.dp)
    ) {

        val state = rememberLazyListState()

        LazyColumn(Modifier.fillMaxSize().padding(end = 12.dp), state) {
            items(50) { x ->
                TextBox("Item #$x", x)
                Spacer(modifier = Modifier.height(5.dp))
            }
        }
        VerticalScrollbar(
            modifier = Modifier.align(Alignment.CenterEnd).fillMaxHeight(),
            adapter = rememberScrollbarAdapter(
                scrollState = state
            )
        )
    }
}

@Composable
fun TextBox(text: String = "Item", height: Int) {
    Box(
        modifier = Modifier.height((30 + height.mod(2) * 50).dp)
            .fillMaxWidth()
            .background(color = Color(0, 0, 0, 20))
            .padding(start = 10.dp),
        contentAlignment = Alignment.CenterStart
    ) {
        Text(text = text)
    }
}
```

When I run this and try to scroll slowly down following happens:

[Screencast 2022-09-21 23:53:54.webm](https://user-images.githubusercontent.com/1088047/191618724-6b53bf12-6121-4e77-b9ba-7d4bf36ae431.webm)

Pay close attention to "shaky" content.

### PR DESCRIPTION
**Pull request by m-sasha**
## Proposed Changes
This mostly re-implements commit
```
793810ba245 25/01/2022, 23:12 Desktop Scrollbar: fix position of a thumb while dragging
```
which broke smooth content scrolling in LazyColumn when not all items are of the same size.
The problem with that commit was that on each mouse-drag event it did (conceptually)
```
dragPosition += dragDelta
contentPosition = dragPosition/scrollScale
```
but when some items are larger than others `scrollScale` can change significantly between the mouse-drag events, so you could have `contentPosition` decrease even when `dragDelta` was always positive.
What we want to do instead is to always add to `contentPosition`:
```
contentPosition += dragDelta/scrollScale
```

The problem that commit was attempting to solve (https://github.com/JetBrains/compose-jb/issues/643) is solved in this commit in a different way:
```
val sliderDelta = 
    (dragPosition + dragDelta).coerceIn(0f, maxScrollPosition) - 
        dragPosition.coerceIn(0f, maxScrollPosition)
contentPosition += sliderDelta/scrollScale
dragPosition += dragDelta
```

Note that this trades-off smooth thumb scrolling for smooth content scrolling, as you can't currently have both (because `LazyScrollbarAdapter.scrollOffset` is discontinuous)

## Testing

Test: Manually on the reproducer submitted in the bug report, and also via unit tests.

Note that the existing unit test was faulty in the sense that it passed even in the presence of the behavior the commit was meant to fix. It was faulty in two ways: the mouse events it was generating were missing the scrollbar thumb, and it was scrolling in the wrong direction for the test.

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-jb/issues/2337

### CODE REVIEW

**Comment by m-sasha** [comment]
Behavior before:

https://user-images.githubusercontent.com/5230206/210507156-20041e4d-6589-4baa-934b-11c0475801bc.mp4

Behavior after:

https://user-images.githubusercontent.com/5230206/210507192-ec1902ef-5fb8-495f-93c2-86287b9423ec.mp4

File: compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt [hunk 1]
```diff
@@ -544,40 +538,45 @@ private class SliderAdapter(
...
+
+    // Stores the unrestricted position during a dragging gesture
+    private var positionDuringDrag = 0f
+
+    // Called when the thumb dragging starts
+    fun onDragStarted() {
+        positionDuringDrag = position
+    }
+
+    // Called on every movement while dragging the thumb
```
Comment (dima-avdeev-jb) [inline compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt]:
```suggestion
    /** Called on every movement while dragging the thumb */
```

It helps to use "Quick Documentation" action to understand function behaviour.

image

File: compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt [hunk 2]
```diff
@@ -544,40 +538,45 @@ private class SliderAdapter(
             return if (extraContentSpace == 0f) 1f else extraScrollbarSpace / extraContentSpace
         }
 
-    /**
-     * A position with cumulative offset, may be out of the container when dragging
-     */
```
Comment (dima-avdeev-jb) [inline compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt:549]: I think this comment may help with better understanding.
Is this logic still correct? (If yes - then better to keep it)

**Review by dima-avdeev-jb — APPROVED** [review]

File: compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt (hunk 1 above)
Comment (m-sasha) [reply compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt]: Changed.

File: compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt (hunk 2 above)
Comment (m-sasha) [reply compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/Scrollbar.desktop.kt:549]: No, this comment and the logic of using it as the cumulative offset was the part of the original commit that I removed.
The position with cumulative (potentially out-of-bounds) offset is now `positionDuringDrag`

**Review by igordmn — APPROVED** [review]

**Pull request merged**